opm/simulators/linalg/bda/BILU0.cpp

```diff
@@ -0,0 +1,381 @@
+        queue->enqueueWriteBuffer(s.Lrows, CL_TRUE, 0, (LMat->Nb + 1) * sizeof(int), LMat->rowPointers);
+        queue->enqueueWriteBuffer(s.Urows, CL_TRUE, 0, (UMat->Nb + 1) * sizeof(int), UMat->rowPointers);
+        queue->enqueueWriteBuffer(s.invDiagVals, CL_TRUE, 0, mat->Nb * sizeof(double) * bs * bs, invDiagVals);
+
+        int *rowsPerColorPrefix = new int[numColors + 1];
+        int prefix = 0;
+        for (int i = 0; i < numColors + 1; ++i) {
+            rowsPerColorPrefix[i] = prefix;
+            prefix += rowsPerColor[i];
+        }
+        queue->enqueueWriteBuffer(s.rowsPerColor, CL_TRUE, 0, (numColors + 1) * sizeof(int), rowsPerColorPrefix);
+        delete[] rowsPerColorPrefix;
+
+        return true;
+    } // end init()
+
+
+    template <unsigned int block_size>
+    bool BILU0<block_size>::create_preconditioner(BlockedMatrix *mat)
+    {
+        const unsigned int bs = block_size;
+        double t1 = 0.0, t2 = 0.0;
+        if (verbosity >= 3){
+            t1 = second();
+        }
+        blocked_reorder_matrix_by_pattern<block_size>(mat, toOrder, fromOrder, rMat);
+        if (verbosity >= 3){
+            t2 = second();
+            std::ostringstream out;
+            out << "BILU0 reorder matrix: " << t2 - t1 << " s";
+            OpmLog::info(out.str());
+        }
+
+        // TODO: remove this copy by replacing inplace ilu decomp by out-of-place ilu decomp
+        if (verbosity >= 3){
+            t1 = second();
+        }
+        memcpy(LUMat->nnzValues, rMat->nnzValues, sizeof(double) * bs * bs * rMat->nnzbs);
+        if (verbosity >= 3){
+            t2 = second();
+            std::ostringstream out;
+            out << "BILU0 memcpy: " << t2 - t1 << " s";
+            OpmLog::info(out.str());
+        }
+
+        int i, j, ij, ik, jk;
+        int iRowStart, iRowEnd, jRowEnd;
+        double pivot[bs * bs];
+
+        int LSize = 0;
+
+        if (verbosity >= 3){
+            t1 = second();
+        }
+        // go through all rows
+        for (i = 0; i < LUMat->Nb; i++) {
+            iRowStart = LUMat->rowPointers[i];
+            iRowEnd = LUMat->rowPointers[i + 1];
+
+            // go through all elements of the row
+            for (ij = iRowStart; ij < iRowEnd; ij++) {
+                j = LUMat->colIndices[ij];
+                // if the element is the diagonal, store the index and go to next row
+                if (j == i) {
+                    diagIndex[i] = ij;
+                    break;
+                }
+                // if an element beyond the diagonal is reach, no diagonal was found
+                // throw an error now. TODO: perform reordering earlier to prevent this
+                if (j > i) {
+                    std::ostringstream out;
+                    out << "BILU0 Error could not find diagonal value in row: " << i;
+                    OpmLog::error(out.str());
+                    return false;
+                }
```

Comment:
This does not really fit here, because the purpose of this for loop is not only to find the diagonal of a row, but also to perform operations using every element before the diagonal of a row.

Reply:
Indeed. Sorry for the noise.